### PR TITLE
Few exception fixes

### PR DESCRIPTION
**Pull request by qparis**

### CODE REVIEW

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java [hunk 1]
```diff
@@ -98,9 +98,11 @@ private void updateOldSelection(ListWidgetSelection<E> oldSelection) {
         Optional.ofNullable(oldSelection).map(ListWidgetSelection::getSelection).ifPresent(selection -> {
             final int oldValueIndex = getControl().getElements().indexOf(selection);
 
-            IconsListElement<E> oldElement = mappedElements.get(oldValueIndex);
+            if (mappedElements.size() > oldValueIndex && oldValueIndex != -1) {
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Why do we need this?
It shouldn't be possible for this case not to happen.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java [hunk 2]
```diff
@@ -146,7 +146,7 @@ public EnginesController(EnginesView enginesView, RepositoryManager repositoryMa
                 this.repositoryManager.addCallbacks(
                         repositoryDTO -> this.enginesManager.fetchAvailableEngines(
                                 repositoryDTO,
-                                engines -> this.populateView(repositoryDTO, engines),
+                                engines -> Platform.runLater(() -> this.populateView(repositoryDTO, engines)),
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java]: Why is this necessary?
I mean `populateView` uses `Platform.runLater` internally therefore it shouldn't be required here

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java]: Hum, not all the content of populateView. showWait() for example

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java (hunk 1 above)
Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: In some cases, it happens... I don't know why

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java (hunk 2 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java]: Yes, but `showWait` shouldn't require `Platform.runLater`.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java (hunk 1 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: I think it is better to locate why this happens.
Like I mentioned the "contract" for `selection` is that it should be always contained inside `getControl().getElements()`.

**Comment by plata** [comment]
The diff is strange.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java (hunk 2 above)
Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java]: Hmmm, interesting.
Then I would suggest to add a `Platform.runLater` around `showWait()`.
This in turn would make it unnecessary to add `Platform.runLater` around `populateView` and `forceViewUpdate`.

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java (hunk 1 above)
Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: The bugs appear randomly so it's hard to reproduce. It think it is because oldValueIndex is sometimes equals to -1

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: The question in my opinion is, why is this the case?
I mean it shouldn't be possible.

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: What if there are no selections?

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Then the `.ifPresent(selection -> {...})` in line 98 won't be called

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java (hunk 2 above)
Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java]: Done

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java (hunk 1 above)
Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Probably a race condition

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: I don't understand how.
This code should be executed by the JavaFX GUI thread.

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: User interaction maybe

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Again I don't understand how.
All user interaction should be processed on the, according to my knowledge, single JavaFX GUI thread.
This in turn should make it impossible for race conditions to occur.

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: capture d ecran 2018-12-30 a 21 04 53

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: capture d ecran 2018-12-30 a 21 06 15

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: In fact, it happens when you select another category

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Did you set any filter settings?
Which applications did the list show?
Do you know the steps to reproduce this?

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: It happens when you select another category

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Ok, the issue is, that a change to the `ListWidgetElement` list doesn't set the selection to null.
The problem here is, that setting the selection to null will still lead to the exception because the `ChangeListener` is triggered.

The only solutions seems to be to add an `if (mappedElements.size() > oldValueIndex && oldValueIndex != -1) {` to the `updateOldSelection()` method like you suggested.
Except that it needs to be added to all three list widget skins.

Still I'm unsure if there isn't a more elegant solution for the problem.

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: I see. Where are the others?

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: - `DetailsListWidgetSkin`
- `CompactListWidgetSkin`
- `IconsListWidgetSkin`

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/icons/skin/IconsListWidgetSkin.java]: Maybe also add a comment describing why the `if` is needed, otherwise it may be accidentally removed later on

File: phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/details/skin/DetailsListWidgetSkin.java [hunk 3]
```diff
@@ -100,7 +100,10 @@ private void updateOldSelection(ListView<DetailsListElement<E>> container, ListW
         Optional.ofNullable(oldSelection).map(ListWidgetSelection::getSelection).ifPresent(selection -> {
             final int oldValueIndex = getControl().getElements().indexOf(selection);
 
-            container.getSelectionModel().clearSelection(oldValueIndex);
+            // In some corner cases (category / filter changes, the current selection can disappear)
+            if (oldValueIndex != -1) {
```
Comment (madoar) [inline phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/details/skin/DetailsListWidgetSkin.java]: I think all three skins require both `mappedElements.size() > oldValueIndex && oldValueIndex != -1`.
I think `container.getSelectionModel().clearSelection(oldValueIndex)` will throw a `NullPointerException` if the index is larger than the number of shown elements.

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/details/skin/DetailsListWidgetSkin.java]: In fact I'm not totally sure. The code is the the same

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/details/skin/DetailsListWidgetSkin.java]: Which code is the same?

Comment (qparis) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/details/skin/DetailsListWidgetSkin.java]: is not*

Comment (madoar) [reply phoenicis-javafx/src/main/java/org/phoenicis/javafx/components/common/widgets/details/skin/DetailsListWidgetSkin.java]: I'm not sure I understand.
The main difference is that the `DetailsListWidget` and the `CompactListWidget` delegate the deselection to the internal `SelectionModel`, while the `IconsListWidget` provides its own logic for the selection and deselection of items.
Both approaches require a valid `oldValueIndex` which is in the interval `[0, mappedElements.size()]` to work correctly.
